### PR TITLE
Remove cuDNN manager

### PR DESCRIPTION
**Pull request by timmoon10**
Previously, we determined whether a layer used a CPU or GPU implementation based on whether it had a null pointer to a cuDNN manger. This approach was deprecated with #250, which started templating layers by device type. The interaction between these two approaches has led to some interesting bugs.

This PR eliminates the cuDNN manager (closing #360). Utilities for CUDA and cuDNN are separated, so only layers that use cuDNN include the cuDNN utility header (I haven't tested, but we might be able to treat cuDNN as a truly optional GPU dependency). Instead of using the `cudnn::cudnn_manager` object, we now access a global cuDNN handle by calling `cudnn::get_handle()`.

Gradient checks come back clean with CPU and GPU runs and I don't see a performance regression on AlexNet on Surface gpgpu.

### CODE REVIEW

File: include/lbann/models/recurrent.hpp [hunk 1]
```diff
@@ -26,6 +26,8 @@
 // recurrent .hpp .cpp - Recurrent neural network models
 ////////////////////////////////////////////////////////////////////////////////
 
+#if 0
```
Comment (naoyam) [inline include/lbann/models/recurrent.hpp]: Why is this disabled?

File: model_zoo/lbann2.cpp [hunk 2]
```diff
@@ -183,51 +183,78 @@ model * build_model_from_prototext(int argc, char **argv, lbann_data::LbannPB &p
...
-      std::cout << std::endl;
-    }
-
     // Save info to file; this includes the complete prototext (with any over-rides
     // from the cmd line) and various other info
     //save_session(comm, argc, argv, pb);
 
-    // Check for cudnn, with user feedback
-    cudnn::cudnn_manager *cudnn = nullptr;
+    // Report useful information
```
Comment (naoyam) [inline model_zoo/lbann2.cpp]: Duplicate of lbann.cpp?

File: model_zoo/lbann_cycgan.cpp [hunk 3]
```diff
@@ -192,51 +192,78 @@ model * build_model_from_prototext(int argc, char **argv, lbann_data::LbannPB &p
...
-      std::cout << std::endl;
-    }
-
     // Save info to file; this includes the complete prototext (with any over-rides
     // from the cmd line) and various other info
     save_session(comm, argc, argv, pb);
 
-    // Check for cudnn, with user feedback
-    cudnn::cudnn_manager *cudnn = nullptr;
+    // Report useful information
```
Comment (naoyam) [inline model_zoo/lbann_cycgan.cpp]: Duplicate?

File: model_zoo/lbann_gan.cpp [hunk 4]
```diff
@@ -185,51 +185,78 @@ model * build_model_from_prototext(int argc, char **argv, lbann_data::LbannPB &p
...
-      std::cout << std::endl;
-    }
-
     // Save info to file; this includes the complete prototext (with any over-rides
     // from the cmd line) and various other info
     //save_session(comm, argc, argv, pb);
 
-    // Check for cudnn, with user feedback
-    cudnn::cudnn_manager *cudnn = nullptr;
+    // Report useful information
```
Comment (naoyam) [inline model_zoo/lbann_gan.cpp]: Duplicate?

File: src/models/recurrent.cpp [hunk 5]
```diff
@@ -24,6 +24,8 @@
 // permissions and limitations under the license.
 ////////////////////////////////////////////////////////////////////////////////
 
+#if 0
```
Comment (naoyam) [inline src/models/recurrent.cpp]: Why is this disabled?

File: src/proto/factories/model_factory.cpp [hunk 6]
```diff
@@ -51,6 +51,7 @@ model* instantiate_model(lbann_comm* comm,
   if (type == "directed_acyclic_graph_model") {
     return new directed_acyclic_graph_model(comm, mini_batch_size, obj, opt);
   }
+#if 0
```
Comment (naoyam) [inline src/proto/factories/model_factory.cpp]: Is this relevant to this PR?

File: include/lbann/models/recurrent.hpp (hunk 1 above)
Comment (timmoon10) [reply include/lbann/models/recurrent.hpp]: I forgot to uncomment, although I think we might want to eliminate the recurrent model. I've opened up issue #467.

File: model_zoo/lbann2.cpp (hunk 2 above)
Comment (timmoon10) [reply model_zoo/lbann2.cpp]: lbann2.cpp, lbann_cycgan.cpp, and lbann_gan.cpp are very similar to lbann.cpp. I know @samadejacobs is working on implementing the GAN stuff within the main LBANN pipeline. We should start thinking about how to replicate the functionality in lbann2.cpp so we can deprecate it.

Comment (samadejacobs) [reply model_zoo/lbann2.cpp]: We should at least keep them for now until may be after SplashApp.

File: include/lbann/models/recurrent.hpp (hunk 1 above)
Comment (naoyam) [reply include/lbann/models/recurrent.hpp]: I think we should uncomment or address #467 first.

File: model_zoo/lbann2.cpp (hunk 2 above)
Comment (timmoon10) [reply model_zoo/lbann2.cpp]: I've opened #468 to discuss this.

File: include/lbann/models/recurrent.hpp (hunk 1 above)
Comment (timmoon10) [reply include/lbann/models/recurrent.hpp]: Recurrent models are removed in #471.

**Comment by naoyam** [comment]
Looks good to me too.
